internal/ext/wasm/wasm.go

```diff
@@ -245,72 +195,56 @@ func (r *Runner) Invoke(ctx context.Context, method string, args any, reply any,
 		return fmt.Errorf("failed to encode codegen request: %w", err)
 	}
 
-	engine := wasmtime.NewEngine()
-	module, err := r.loadModule(ctx, engine)
+	cacheDir, err := cache.PluginsDir()
 	if err != nil {
-		return fmt.Errorf("loadModule: %w", err)
+		return err
 	}
 
-	linker := wasmtime.NewLinker(engine)
-	if err := linker.DefineWasi(); err != nil {
+	cache, err := wazero.NewCompilationCacheWithDir(filepath.Join(cacheDir, "wazero"))
+	if err != nil {
 		return err
 	}
 
-	dir, err := os.MkdirTemp(os.Getenv("SQLCTMPDIR"), "out")
+	wasmBytes, err := r.loadBytes(ctx)
 	if err != nil {
-		return fmt.Errorf("temp dir: %w", err)
+		return fmt.Errorf("loadModule: %w", err)
 	}
 
-	defer os.RemoveAll(dir)
-	stdinPath := filepath.Join(dir, "stdin")
-	stderrPath := filepath.Join(dir, "stderr")
-	stdoutPath := filepath.Join(dir, "stdout")
-
-	if err := os.WriteFile(stdinPath, stdinBlob, 0755); err != nil {
-		return fmt.Errorf("write file: %w", err)
-	}
+	config := wazero.NewRuntimeConfig().WithCompilationCache(cache)
+	rt := wazero.NewRuntimeWithConfig(ctx, config)
+	defer rt.Close(ctx)
 
-	// Configure WASI imports to write stdout into a file.
-	wasiConfig := wasmtime.NewWasiConfig()
-	wasiConfig.SetArgv([]string{"plugin.wasm", method})
-	wasiConfig.SetStdinFile(stdinPath)
-	wasiConfig.SetStdoutFile(stdoutPath)
-	wasiConfig.SetStderrFile(stderrPath)
+	// TODO: Handle error
+	wasi_snapshot_preview1.MustInstantiate(ctx, rt)
 
-	keys := []string{"SQLC_VERSION"}
-	vals := []string{info.Version}
-	for _, key := range r.Env {
-		keys = append(keys, key)
-		vals = append(vals, os.Getenv(key))
+	// Compile the Wasm binary once so that we can skip the entire compilation time during instantiation.
+	mod, err := rt.CompileModule(ctx, wasmBytes)
```

Comment:
If it's possible, it would be nice to rejigger to scope this to `Runner`, possibly with some `map[/* wasm url */ string]wazero.CompiledModule`. The compilation cache is good to reuse across executions of the sqlc process itself, but it's also good to only compile once per wasm within a process if possible, the cache key computation isn't trivial. Though if the latter doesn't happen that much maybe it doesn't matter